packages/lit-html/src/directives/async-helpers.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * Holds a reference to an instance that can be disconnected and reconnected,
+ * so that a closure over the ref (e.g. in a then function to a promise) does
+ * not strongly hold a ref to the instance. Approximates a WeakRef but must
+ * be manually connected & disconnected to the backing instance.
+ */
+export class PseudoWeakRef<T> {
+  private _ref?: T;
+  constructor(ref: T) {
+    this._ref = ref;
+  }
+  /**
+   * Disassociates the ref with the backing instance.
+   */
+  disconnect() {
+    this._ref = undefined;
+  }
+  /**
+   * Reassociates the ref with the backing instance.
+   */
+  reconnect(ref: T) {
+    this._ref = ref;
+  }
+  /**
+   * Retrieves the backing instance (will be undefined when disconnected)
+   */
+  deref() {
+    return this._ref;
+  }
+}
+
+/**
+ * A helper to pause and resume waiting on a condition in an async function
+ */
+export class Pauser {
```

Comment:
Perhaps bikeshedding, but it seems slightly better if this is called a `Player` and has `play` and `pause` methods. It could then have an `isPaused()` method which returns a promise that's resolved when `play` is called. Could that promise also just encapsulate the for loop that's currently used with `get`? The API seems less awkward that way.

Reply:
Or maybe `Runner`?

Reply:
I started to implement it that way, but the final falsey check on `pauser.get()` must be synchronous to resuming, and you can't encapsulate that in an async function because the outside `await` adds a microtask between which `pauser.get()` might return truthy again and thus would be wrong.